unitary/alpha/quantum_world.py

```diff
@@ -66,7 +110,8 @@ def undo_last_effect(self):
         Note that pop() is considered to be an effect for the purposes
         of this call.
         """
-        self.circuit, self.post_selection = self.effect_history.pop()
+        if self.effect_history:
```

Comment:
Throwing an exception seems better if there is no last effect.

Reply:
Done.